devices/Sim7080/SimController.cs

```diff
@@ -0,0 +1,261 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.IO.Ports;
+using System.Text.RegularExpressions;
+using System.Threading;
+
+namespace IoT.Device.Sim7080
+{
+    internal static class SimController
+    {
+        /// <summary>
+        /// Set the network system mode for wireless and cellular network communication service
+        /// </summary>
+        /// <param name="serialPort">The UART <see cref="SerialPort"/> for communication with the modem</param>
+        /// <param name="systemMode">The desired <see cref="SystemMode"/></param>
+        /// <param name="enableReporting">Report the network system mode information</param>
+        public static void SetSystemMode(SerialPort serialPort, SystemMode systemMode, bool enableReporting)
+        {
+            try
+            {
+                var reportingEnabled = (enableReporting) ? 1 : 0;
+
+                ExecuteCommand(serialPort, $"AT+CNSMOD={reportingEnabled},{(int)systemMode}");
+            }
+            catch (Exception exception)
+            {
+                Debug.WriteLine(exception.Message);
+            }
+        }
+
+        /// <summary>
+        /// Connect to the cellular network
+        /// </summary>
+        /// <param name="serialPort">The UART <see cref="SerialPort"/> for communication with the modem</param>
+        /// <param name="apn">Cellular network access point name</param>
+        /// <returns><see cref="ConnectionStatus"/></returns>
+        public static ConnectionStatus NetworkConnect(SerialPort serialPort, string apn)
+        {
+            try
+            {
+                //Read Signal Quality
+                ExecuteCommand(serialPort, "AT+CSQ");
+
+                //Return current Operator
+                ExecuteCommand(serialPort, "AT+COPS?");
+
+                //Get Network APN in CAT-M or NB-IoT
+                ExecuteCommand(serialPort, "AT+CGNAPN");
+
+                //Define PDP Context, saves APN
+                ExecuteCommand(serialPort, $"AT+CGDCONT=1,\"IP\",\"{apn}\"");
+
+                //App Network Active, assign IP
+                ExecuteCommand(serialPort, "AT+CNACT=0,2");
+
+                //Read IP
+                ExecuteCommand(serialPort, "AT+CNACT?");
```

Comment:
what if some of those commands are not sucessfull? What if you don't get a proper IP address?
In your logic, you'll set it up to connected but later in the code, you're checking that you're having a valid address. So I guess there is a bit of logic missing here

Reply:
Each separate command creates an acknowledgement. The IP address can be derived from the acknowledgement message, which is handeld by the ReadReponse() method. If the IP is 0.0.0.0 than ConnectionStatus will be Disconnected.

The retry will try 3 times; after 3 times it will fail. In you procedure you should therefor always check wether the previous step was successfully. 

Unfortunatly there is no detailed message explaining when something goes wrong.

So tried to make it robust using retries and the ConnectionStatus.

Reply:
ok, get the logic, works for me. Now, see one of the other comments about exposing somehow the status there as well

Reply:
The actual ACK is:

+CNACT: 0,1,"10.94.36.44" +CNACT: 1,0,"0.0.0.0" +CNACT: 2,0,"0.0.0.0" +CNACT: 3,0,"0.0.0.0"

Reply:
Thanks. I trust you're doing the right thing! Thanks for providing.